### PR TITLE
Enabling Transformer fast path for not batch_first (MHA, TE, TEL)

### PR DESCRIPTION
**Pull request by mikekgfb**
Summary: The fast path for the `forward()` method in `MultiheadAttention`, `TE`, `TEL` only accepted `batch_first = True`. This diff enables fast path for `batch_first=False` as well.

Differential Revision: D48095703

cc @mrshenli @pritamdamania87 @zhaojuanmao @satgera @rohan-varma @gqchen @aazzolini @osalpekar @jiayisuse @H-Huang @kwen2501 @awgu @penguinwu @fegin @XilunWu @wanchaol @fduwjj @wz337 @tianyu-l @wconstab @yf225 @kiukchung @d4l3k @lucasllc

### CODE REVIEW

**Comment by mikekgfb** [comment]
@wconstab any suggestion how to handle the FSDP fail?  I think (i.e., conjecture) the underlying cause is a numerical difference between fastpath vs. standard execution -- I think this is becawe avoided running into this so far because this test ran with batch_first=False, and until this diff we used the fastpath with batch_first=True only. (fastpath is only on for inference with no_grad(), so eval() vs eval+no_grad() exercises two different paths)

cc: @rohan-varma @awgu @mrshenli @drisspg

**Comment by wconstab** [comment]
missed this comment before, but it sounds like you've found out that its an issue of eval vs train accuracy for this model, is that right?

**Comment by mikekgfb** [comment]
> missed this comment before, but it sounds like you've found out that its an issue of eval vs train accuracy for this model, is that right?

The underlying issue is that model.eval() vs model.eval() with no_grad() triggers different computational kernels.  The only way to control this in a sane way is via a backend context manager for choosing between these backends.  I added this in #107014

**Comment by mikekgfb** [comment]
> missed this comment before, but it sounds like you've found out that its an issue of eval vs train accuracy for this model, is that right?

Yep - the way that "validation" is performed is that the test runs with train mode, and with eval/no_grad -- the latter triggers fastpath with the present bathc, but because we;re looking at different implementations, we can't expect bit-exact answers.

One possible solution is a context manager that gives more control over the kernel chosen, similar for what we do for SDPA - #107163 is an implementation of this context manager for backend selection
